### PR TITLE
CSHA256 compression function

### LINKED ISSUE #814: SHA256 compression function doesn't get called when not enough data is passed
In both of our SHA256 compression function implementations:

- https://github.com/zcash/zcash/pull/804/files#diff-cf563899256ae1ec7ee78b61806e40caL153
- https://github.com/Electric-Coin-Company/libzerocash/blob/zc.latest.0/libzerocash/utils/sha256.cpp#L99

If you pass in less than 64 bytes, and use them to get the compression function (rather than the full hash) as in this test pattern https://github.com/zcash/zcash/pull/804/files#diff-e872c9211727e681e7485031acee832aR56 then the compression function doesn't actually get called at all!

### PR DESCRIPTION
**Pull request by ebfull**
This implements our SHA256 compression function in Bitcoin's own CSHA256 and also removes libzerocash's SHA256 implementation.

Closes #814

### CODE REVIEW

File: src/test/sha256compress_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+			                digest.GetHex());
+	}
+
+	{
+		unsigned char preimage[63] = {};
+		CSHA256 hasher;
+		hasher.Write(&preimage[0], 63);
+		uint256 digest;
+		BOOST_CHECK_THROW(hasher.FinalizeNoPadding(digest.begin()), std::length_error);
+	}
```
Comment (defuse) [inline src/test/sha256compress_tests.cpp]: We need two more tests:

1. Passing in 65 bytes.
2. Passing in 64 bytes, one byte at a time, each one with its own call to Write.

File: src/zerocash/tests/utilTest.cpp [hunk 2]
```diff
@@ -3,17 +3,39 @@
...
+                                                               'a', 'b', 'c', 'd',
+                                                               'a', 'b', 'c', 'd',
+                                                               'a', 'b', 'c', 'd',
+                                                               'a', 'b', 'c', 'd',
+                                                               'a', 'b', 'c', 'd',
+                                                               'a', 'b', 'c', 'd',
+                                                               'a', 'b', 'c', 'd'
+                                                             };
 /* This is the SHA256 hash of "abc" according to the modified implementation of
  * SHA256 included in libzerocash. */
```
Comment (defuse) [inline src/zerocash/tests/utilTest.cpp]: This comment is now incorrect.

File: src/zerocash/utils/sha256.h [hunk 3]
```diff
@@ -1,32 +0,0 @@
...
-	uint32_t state[8];
-} SHA256_CTX_mod;
-
-/*********************** FUNCTION DECLARATIONS **********************/
-void sha256_init(SHA256_CTX_mod *ctx);
-void sha256_update(SHA256_CTX_mod *ctx, const uint8_t data[], size_t len);
-void sha256_length_padding(SHA256_CTX_mod *ctx);
-void sha256_final_no_padding(SHA256_CTX_mod *ctx, uint8_t hash[]);
-
-#endif   // SHA256H_H
```
Comment (defuse) [inline src/zerocash/utils/sha256.h]: I was happy that we were testing both implementations against each other, but there's issue #813 about doing that with different-language implementations which would be even better.

Comment (defuse) [inline src/zerocash/utils/sha256.h]: (And testing the implementations against each other didn't catch #814 so we still need good review.)

File: src/zerocash/utils/util.cpp [hunk 4]
```diff
@@ -203,68 +203,65 @@ void concatenateVectors(const std::vector<unsigned char>& A, const std::vector<u
...
+    hasher.FinalizeNoPadding(hash);
 }
 
-void sha256(SHA256_CTX_mod* ctx256, const unsigned char* input, unsigned char* hash, int len) {
-	sha256_init(ctx256);
-	sha256_update(ctx256, input, len);
-	sha256_final_no_padding(ctx256, hash);
+void sha256(CSHA256& hasher, const unsigned char* input, unsigned char* hash, int len) {
+	hasher.Write(input, len);
+    hasher.FinalizeNoPadding(hash);
```
Comment (defuse) [inline src/zerocash/utils/util.cpp]: The old behavior was to re-initialize the context. The new behavior doesn't do that. Does that matter?

**Comment by defuse** [comment]
ACK after those tests are added.

File: src/zerocash/utils/util.cpp (hunk 4 above)
Comment (ebfull) [inline src/zerocash/utils/util.cpp]: I have it making a new hasher now.

**Comment by defuse** [comment]
ACK

**Comment by defuse** [comment]
@zkbot r+

**Comment by defuse** [comment]
The test failed because sourceforge isn't serving this URL: `http://sourceforge.net/projects/boost/files/boost/1.57.0/boost_1_57_0.tar.bz2`. If you put that in your browser right now you'll see a 404 page, but the HTTP status code is 200 OK, not 404. So the `wget` in the depends system succeeds and it doesn't fall back to using our mirror.

**Comment by defuse** [comment]
@zkbot retry

**Comment by defuse** [comment]
@zkbot r+

**Pull request merged**